File: Path Sum II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> >result;
    vector<int> oneResult;
    
    vector<vector<int> > pathSum(TreeNode *root, int sum) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        result.clear();
        oneResult.clear();
        if(root)
            checkPathSum(root, sum, false);
        return result;
    }
    
    void checkPathSum(TreeNode *root, int sum, bool isLeaf)
    {
   
        oneResult.push_back(root->val);
        int val = sum - root->val;
        if(root->left==NULL&&root->right==NULL)
        {
            if(val==0)
                result.push_back(oneResult);
        }
        else
        {
            if(root->left)
                checkPathSum(root->left, val, isLeaf);
            if(root->right)
                checkPathSum(root->right, val, isLeaf);
        }
        
        
        oneResult.pop_back();
        
    }
};

File: Path Sum II.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ArrayList<ArrayList<Integer>> pathSum(TreeNode root, int sum) {
        // Start typing your Java solution below
        // DO NOT write main() function
        ArrayList<ArrayList<Integer>> results = new ArrayList<ArrayList<Integer>>();
        ArrayList<Integer> ret = new ArrayList<Integer>();
        if(root != null)
        {
            getSum(results, ret, root, 0, sum);
        }
        return results;
    }
    
    private void getSum(ArrayList<ArrayList<Integer>> results, ArrayList<Integer> ret, TreeNode root, int val, int sum)
    {
        val += root.val;
        ret.add(root.val);
        if(root.left==null && root.right==null && val==sum)
        {
            results.add((ArrayList<Integer>)ret.clone());
        }
        if(root.left != null)
        {
            getSum(results, ret, root.left, val, sum);
        }
        if(root.right != null)
        {
            getSum(results, ret, root.right, val, sum);
        }
        int lastIndex = ret.size() - 1;
        ret.remove(lastIndex);
    }
}